tools/chainsync-loadtest/src/concurrency/once.rs

```diff
@@ -0,0 +1,58 @@
+use std::future::Future;
+use std::sync::RwLock;
+
+fn is_send<T: Send>() {}
+fn is_sync<T: Sync>() {}
+
+#[allow(dead_code)]
+fn test<T: Clone + Send + Sync>() {
+    is_send::<Once<T>>();
+    is_sync::<Once<T>>();
+}
+
+// Once is a synchronization primitive, which stores a single value.
+// This value can be set at most once, and multiple consumers are
+// allowed to wait for that value.
+pub struct Once<T: Clone + Send + Sync> {
+    value: RwLock<Option<T>>,
+    notify: tokio::sync::Notify,
+}
+
+impl<T: Clone + Send + Sync> Once<T> {
+    pub fn new() -> Once<T> {
+        return Once { value: RwLock::new(None), notify: tokio::sync::Notify::new() };
+    }
+
+    // set() sets the value of Once to x and returns true.
+    // If the value was already set, it just returns false.
+    pub fn set(self: &Self, x: T) -> bool {
```

Comment:
why?

Reply:
To give a brief answer, because the return value here indicates a success/failure of the requested operation, and `Result` does a much better job at documenting this than `bool`, both to the user and the compiler.

Reply:
ack, done.